### PR TITLE
NIFI-2566: Refactoring to improve robustness of cluster

### PR DESCRIPTION
**Pull request by markap14**

### CODE REVIEW

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/cluster/ClusterProtocolHeartbeater.java [hunk 1]
```diff
@@ -18,102 +18,80 @@
...
+
+        final List<NodeConnectionStatus> updatedStatuses = responseMessage.getUpdatedNodeStatuses();
+        if (updatedStatuses != null) {
+            for (final NodeConnectionStatus updatedStatus : updatedStatuses) {
+                final NodeIdentifier nodeId = updatedStatus.getNodeIdentifier();
+                final Long updateId = updateIdMap.get(nodeId);
+
+                final boolean updated = clusterCoordinator.resetNodeStatus(updatedStatus, updateId == null ? -1L : updateId);
+                if (updated) {
+                    // TODO - REMOVE THE ASTERISKS
```
Comment (JPercivall) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/cluster/ClusterProtocolHeartbeater.java]: Lingering Todo, can it be removed?

**Comment by JPercivall** [comment]
@markap14, I'm having trouble running a standalone instance with this PR. I get the at I can't start the FlowController due to (very long message, took the end):

`Caused by: java.lang.IllegalStateException: The 'nifi.zookeeper.connect.string' property is not set in nifi.properties
        at org.apache.nifi.controller.cluster.ZooKeeperClientConfig.createConfig(ZooKeeperClientConfig.java:76) ~[nifi-framework-core-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
        at org.apache.nifi.controller.leader.election.CuratorLeaderElectionManager.<init>(CuratorLeaderElectionManager.java:61) ~[nifi-framework-core-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[na:1.8.0_74]
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[na:1.8.0_74]
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[na:1.8.0_74]
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423) ~[na:1.8.0_74]
        at org.springframework.beans.BeanUtils.instantiateClass(BeanUtils.java:147) ~[spring-beans-4.2.4.RELEASE.jar:4.2.4.RELEASE]
        ... 52 common frames omitted`

**Comment by markap14** [comment]
@JPercivall I pushed an update that should address this issue.

**Comment by JPercivall** [comment]
+1

Visually verified code, and did a contrib check build rebased to master. In a 3 node secure cluster did all sorts of tests involving connecting, disconnecting, restarting, and starting/stopping components. All behaved as expected. Thanks @markap14, I will merge it in.

**Pull request closed without merging**